runtime/lua/vim/treesitter/query.lua

```diff
@@ -538,17 +585,19 @@ end
 --- Adds a new directive to be used in queries
 ---
 --- Handlers can set match level data by setting directly on the
---- metadata object `metadata.key = value`, additionally, handlers
+--- metadata object `metadata.key = value`. Additionally, handlers
 --- can set node level data by using the capture id on the
 --- metadata table `metadata[capture_id].key = value`
 ---
 ---@param name string Name of the directive, without leading #
----@param handler function(match:table<string,TSNode>, pattern:string, bufnr:integer, predicate:string[], metadata:table)
----   - match: see |treesitter-query|
----      - node-level data are accessible via `match[capture_id]`
----   - pattern: see |treesitter-query|
+---@param handler function(match:table<integer,TSNode[]>, pattern:integer, bufnr:integer, predicate:string[], metadata:table)
+---   - match: A table mapping capture IDs to a list of captured nodes
+---   - pattern: the index of the matching pattern in the query file
 ---   - predicate: list of strings containing the full directive being called, e.g.
 ---     `(node (#set! conceal "-"))` would get the predicate `{ "#set!", "conceal", "-" }`
+---   - any: (only used for predicates) true if this predicate should match if ANY of the captured
+---       nodes satisfy the predicate; otherwise, the predicate should match if ALL of the captured
+---       nodes satsify the predicate.
```

Comment:
I think this is the implementation detail that should be encapsulated into the predicate/directive handler, so I suggest we do not have this parameter to add_directive and add_predicate. The handler functions would get the full `match` mapping of `table<integer,TSNode[]>` rather than `table<integer,TSNode>`, so whether to behave like `any` or `all` is a responsibility of handler functions. With that (i.e. not having the any parameter), we don't need to change the interface of predicates/directives (other than the breaking change on the type of `table`), and the interface of predicates/directives could become more general.

Reply:
Do we actually need this for anything? If not I'd suggest we'd leave this out for now.

Reply:
Need what? We don't need `any` for directives, only predicates, but the docstring on `add_predicate` just says "see `add_directive()`", so I put the docs here.

Reply:
Do we need `any` for any of our predicates? This deviates from upstream which I'd like to avoid.

Reply:
Ok, I was just struggling to review this in the github UI, I think I get it now. @wookayin does make a good point though, we can probably do this without changing the predicate interface.

Reply:
> Yes we do, at least kind of, because now predicates have to implement the "all" vs "any" semantics themselves.

Which I'd argue is correct, since their semantics (in the multi-capture case) depends on what comes _after_ the `any` etc.

Reply:
Let me elaborate -- sorry for some mistakes in the description and if it wasn't clear enough.

> We are not adding a parameter to add_directive or add_predicate,

Yes you're right. I was talking about the new `any` parameter, which is actually being added to predicate handler, not to `add_{directive,predicate}` --- I said it wrong by mistake. What I was suggesting is *not* to add this new parameter to *predicate handlers*; Instead, leave the semantics (like `any-` or `all`-) decided by the predicate handler itself, i.e. "*they have to implement the all vs any semantics themselves*". It would look like:

```lua
predicate_handlers['eq?'] = function(match, pattern, source, predicate)
  return eq_impl(match, pattern, source, predicate, false) -- work like `all?`
end,
predicate_handlers['any-eq?'] = function(match, pattern, source, predicate)
  return eq_impl(match, pattern, source, predicate, true) -- work like `not?`
end
```

It'd be an contrived and hypothetical example, but one could think of a predicate that is neither `all-` or `any-`; say `at-least-three` or `exclusive-or` (if `#match == 2`), for instance. For such predicates the parameter `any` would just be an unused parameter to ignore. This idea would also apply to the directives (as @amaanq suggested in https://github.com/neovim/neovim/pull/24738#issuecomment-1926284050). We can also think of predicates/directives that would *count*; or an user/plugin might write an arbitrary complex logic to deal with multiple nodes captured.

This design also implies that we will not `any-` have be a general logic quantifier prefix unlike `not-` (which simply negates the return value); instead, we allow `any-` for only certain predicates that we know their implementation details. I think this is also closer to [what upstream tree-sitter does](https://github.com/tree-sitter/tree-sitter/pull/2532/files#diff-2af4e27edd56f326a108d99d345842f5f477fe2f933ca50cdd7c4eacfbffdf65).

Reply:
Ok that makes much more sense, thank you for explaining.

I think I'm ok with that route, it does mean we have to implement some predicates effectively twice (the logic between `eq?` and `any-eq?` is basically the same), but that is not a big deal (or we can just make a helper implementation, as you hinted).

Reply:
I've just force pushed this change.